### ISSUE TITLE
Lint for `&String::into(): String` on final use of input string

### ISSUE THREAD

**Issue by pnkfelix** [rust-lang/rust#61994]
PR #59825 added a `impl From<&String> for String`.

While perusing the `rustc` source, I am seeing instances of `String` values being constructed and then passed via-reference into a function that wants a `String`.

It might be worthwhile to try to make a lint that detects the last use of a `String` that is then being used as an argument to `&` which is then itself passed to this conversion method (which effectively hides the clone taking place.

**Comment by awestlake87** [comment]
Yeah, I'm not sure I like that change. I always liked that potentially expensive operations like clone() were explicit and searchable. Also, I usually think of `.into()` and `::from` as a move, not a copy, since it consumes the callee. 

In C++, it's extremely easy to make copies accidentally and while `impl From<&String> for String` might be reasonably harmless, I feel like an `impl <T: Clone> From<&T> for T` kinda pushes Rust too far in that direction.

If that feature is here to stay, then I would probably always enable this lint as an error in my projects
